### PR TITLE
HIVE-24878: ClassNotFound exception for function replication

### PR DESCRIPTION
**Pull request by ArkoSharma**

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/load/message/CreateFunctionHandler.java [hunk 1]
```diff
@@ -206,7 +206,7 @@ ResourceUri destinationResourceUri(ResourceUri resourceUri)
         return ReplCopyTask.getLoadCopyTask(metadata.getReplicationSpec(), new Path(sourceUri), dest, context.hiveConf,
                 context.getDumpDirectory(), context.getMetricCollector());
       } else {
-        return TaskFactory.get(new CopyWork(new Path(sourceUri), dest, true, false,
+        return TaskFactory.get(new CopyWork(new Path(sourceUri), dest.getParent(), true, false,
```
Comment (pkumarsinha) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/load/message/CreateFunctionHandler.java:209]: Just wondering how was this working earlier.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcrossInstances.java [hunk 2]
```diff
@@ -2237,4 +2242,27 @@ private void assertExternalFileInfo(List<String> expected, String dumplocation,
...
+    //correct location of jars on target is functionRoot/dbName/funcName/nanoTs/jarFile
+    FileStatus[] fileStatuses = replica.miniDFSCluster.getFileSystem()
+            .globStatus(
+               new Path(replica.functionsRoot + "/" + replicatedDbName.toLowerCase() + "/" +
+                       functionName.toLowerCase() + "/*/*"), path -> path.toString().endsWith("jar")
+            );
+    assertEquals(fileStatuses.length, expectedJars.size());
+    List<String> jars = Arrays.stream(fileStatuses).map(f -> {
+      String[] splits = f.getPath().toString().split("/");
+      return splits[splits.length - 1];
```
Comment (pkumarsinha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcrossInstances.java:2256]: Do you need just file name from the path? Wouldn't just return f.getPath().getName() be fine

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcrossInstances.java [hunk 3]
```diff
@@ -2237,4 +2242,27 @@ private void assertExternalFileInfo(List<String> expected, String dumplocation,
...
+    }).collect(Collectors.toList());
+    assertThat(jars, containsInAnyOrder(expectedJars.toArray()));
+
+    //confirm no jars in nested directories
+    FileStatus[] nestedFileStatuses = replica.miniDFSCluster.getFileSystem()
+            .globStatus(
+                    new Path(replica.functionsRoot + "/" + replicatedDbName.toLowerCase() + "/" + functionName + "/*/*/*")
+                    ,path -> path.toString().endsWith("jar")
+            );
+    assertEquals(nestedFileStatuses.length, 0);
```
Comment (pkumarsinha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcrossInstances.java:2266]: There should not be any directory itself right?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcrossInstances.java [hunk 4]
```diff
@@ -2237,4 +2242,27 @@ private void assertExternalFileInfo(List<String> expected, String dumplocation,
...
+            .globStatus(
+               new Path(replica.functionsRoot + "/" + replicatedDbName.toLowerCase() + "/" +
+                       functionName.toLowerCase() + "/*/*"), path -> path.toString().endsWith("jar")
+            );
+    assertEquals(fileStatuses.length, expectedJars.size());
+    List<String> jars = Arrays.stream(fileStatuses).map(f -> {
+      String[] splits = f.getPath().toString().split("/");
+      return splits[splits.length - 1];
+    }).collect(Collectors.toList());
+    assertThat(jars, containsInAnyOrder(expectedJars.toArray()));
```
Comment (pkumarsinha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcrossInstances.java:2258]: Does this also do a match on size

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcrossInstances.java [hunk 5]
```diff
@@ -2237,4 +2242,27 @@ private void assertExternalFileInfo(List<String> expected, String dumplocation,
     }
     ReplicationTestUtils.assertExternalFileInfo(warehouseInstance, expected, externalTableInfoFile);
   }
+
+  private void assertFunctionJarsOnTarget(String functionName, List<String> expectedJars) throws IOException {
+    //correct location of jars on target is functionRoot/dbName/funcName/nanoTs/jarFile
+    FileStatus[] fileStatuses = replica.miniDFSCluster.getFileSystem()
+            .globStatus(
+               new Path(replica.functionsRoot + "/" + replicatedDbName.toLowerCase() + "/" +
+                       functionName.toLowerCase() + "/*/*"), path -> path.toString().endsWith("jar")
```
Comment (pkumarsinha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcrossInstances.java:2251]: endsWith(".jar")?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcrossInstances.java (hunk 2 above)
Comment (ArkoSharma) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcrossInstances.java:2256]: That will work, will make this change.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcrossInstances.java (hunk 3 above)
Comment (ArkoSharma) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcrossInstances.java:2266]: Yes, i think a better check after matching the jar names will just be to confirm none of the files are directories. That will also be sufficient to test there are no nested directories. Will make this change.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcrossInstances.java (hunk 4 above)
Comment (ArkoSharma) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcrossInstances.java:2258]: This is done in a preceding line : assertEquals(fileStatuses.length, expectedJars.size());

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcrossInstances.java (hunk 5 above)
Comment (ArkoSharma) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcrossInstances.java:2251]: This check is actually not required, removing it.

**Review by pkumarsinha — APPROVED** [review]
+1 LGTM

**Pull request merged**